# Project 1: SAT & ACT Analysis

The first markdown cell in a notebook is a great place to provide an overview of your entire project. You will likely want to at least state your

## Problem Statement

as well as an

## Executive Summary

If you want to, it's great to use relative links to direct your audience to various sections of a notebook. **HERE'S A DEMONSTRATION WITH THE CURRENT SECTION HEADERS**:

### Contents:
- [2017 Data Import & Cleaning](#Data-Import-and-Cleaning)
- [2018 Data Import and Cleaning](#2018-Data-Import-and-Cleaning)
- [Exploratory Data Analysis](#Exploratory-Data-Analysis)
- [Data Visualization](#Visualize-the-data)
- [Descriptive and Inferential Statistics](#Descriptive-and-Inferential-Statistics)
- [Outside Research](#Outside-Research)
- [Conclusions and Recommendations](#Conclusions-and-Recommendations)

**If you combine your problem statement, executive summary, data dictionary, and conclusions/recommendations, you have an amazing README.md file that quickly aligns your audience to the contents of your project.** Don't forget to cite your data sources!

*All libraries used should be added here*

In [61]:
#Imports:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
import scipy.stats as stats

%matplotlib inline
%config InlineBackend.figure_format = 'retina'


## 2017 Data Import and Cleaning

#### 1. Read In SAT & ACT  Data

Read in the `sat_2017.csv` and `act_2017.csv` files and assign them to appropriately named pandas dataframes.

In [62]:
#couldnt figure out the relative path when i started.
sat = pd.read_csv(r'/Users/owner/Desktop/gclone/project_1/project_1/project-1/data/sat_2017.csv')
sat.head(21)

State Participation  Evidence-Based Reading and Writing  \
0                Alabama            5%                                 593   
1                 Alaska           38%                                 547   
2                Arizona           30%                                 563   
3               Arkansas            3%                                 614   
4             California           53%                                 531   
5               Colorado           11%                                 606   
6            Connecticut          100%                                 530   
7               Delaware          100%                                 503   
8   District of Columbia          100%                                 482   
9                Florida           83%                                 520   
10               Georgia           61%                                 535   
11                Hawaii           55%                                 544   
12                 Idaho           93%                                 513   
13              Illinois            9%                                 559   
14               Indiana           63%                                 542   
15                  Iowa            2%                                 641   
16                Kansas            4%                                 632   
17              Kentucky            4%                                 631   
18             Louisiana            4%                                 611   
19                 Maine           95%                                 513   
20              Maryland           69%                                 536   

    Math  Total  
0    572   1165  
1    533   1080  
2    553   1116  
3    594   1208  
4    524   1055  
5    595   1201  
6    512   1041  
7    492    996  
8    468    950  
9    497   1017  
10   515   1050  
11   541   1085  
12   493   1005  
13   556   1115  
14   532   1074  
15   635   1275  
16   628   1260  
17   616   1247  
18   586   1198  
19   499   1012  
20    52   1060

In [63]:
act = pd.read_csv(r'/Users/owner/Desktop/gclone/project_1/project_1/project-1/data/act_2017.csv')
act.head()

State Participation  English  Math  Reading  Science Composite
0  National           60%     20.3  20.7     21.4     21.0      21.0
1   Alabama          100%     18.9  18.4     19.7     19.4      19.2
2    Alaska           65%     18.7  19.8     20.4     19.9      19.8
3   Arizona           62%     18.6  19.8     20.1     19.8      19.7
4  Arkansas          100%     18.9  19.0     19.7     19.5      19.4

#### 2. Display Data

Print the first 10 rows of each dataframe to your jupyter notebook

In [64]:
sat.head(10)

State Participation  Evidence-Based Reading and Writing  \
0               Alabama            5%                                 593   
1                Alaska           38%                                 547   
2               Arizona           30%                                 563   
3              Arkansas            3%                                 614   
4            California           53%                                 531   
5              Colorado           11%                                 606   
6           Connecticut          100%                                 530   
7              Delaware          100%                                 503   
8  District of Columbia          100%                                 482   
9               Florida           83%                                 520   

   Math  Total  
0   572   1165  
1   533   1080  
2   553   1116  
3   594   1208  
4   524   1055  
5   595   1201  
6   512   1041  
7   492    996  
8   468    950  
9   497   1017

In [65]:
act.head(10)

State Participation  English  Math  Reading  Science  \
0              National           60%     20.3  20.7     21.4     21.0   
1               Alabama          100%     18.9  18.4     19.7     19.4   
2                Alaska           65%     18.7  19.8     20.4     19.9   
3               Arizona           62%     18.6  19.8     20.1     19.8   
4              Arkansas          100%     18.9  19.0     19.7     19.5   
5            California           31%     22.5  22.7     23.1     22.2   
6              Colorado          100%     20.1  20.3     21.2     20.9   
7           Connecticut           31%     25.5  24.6     25.6     24.6   
8              Delaware           18%     24.1  23.4     24.8     23.6   
9  District of Columbia           32%     24.4  23.5     24.9     23.5   

  Composite  
0      21.0  
1      19.2  
2      19.8  
3      19.7  
4      19.4  
5      22.8  
6      20.8  
7      25.2  
8      24.1  
9      24.2

#### 3. Verbally Describe Data

Take your time looking through the data and thoroughly describe the data in the markdown cell below. 

Answer:
SAT: 
    -I see no NaN values. (.isna)
    -It is a DataFrame (type())
    -No duplicate values (.duplicate)
    -Wyoming has the max values in participation, reading-writing, and math (.max)
    -Alabama has the lowest scores, but not the lowest participation rate. (.min) 
    -shape of 51 rows and 5 columns (.shape) 
    -Missing total Averages
ACT: 
    -I see no NaN values. (.isna)
    -It is a DataFrame (type())
    -No duplicate values (.duplicate)
    -Wyoming has the max values in participation, reading-writing, and math (.max)
    -Alabama has the lowest scores, but not the lowest participation rate. Alabama's participation is much higherhere      than SAT participation  (.min) 
    - ACT gives us national AVG. 

#### 4a. Does the data look complete? 

Answer: yes, some values need to be corrected. 

#### 4b. Are there any obvious issues with the observations?
        -there were two errors that i didnt realize till I plotted
#### 4c.Not so obvious to me:
        -Maryland has a 52 isntead of 524, fixed after i merged my dataframes
        -Washington D.C vs . Distric of columbia 

**What is the minimum *possible* value for each test/subtest? What is the maximum *possible* value?**

Consider comparing any questionable values to the sources of your data:
- [SAT](https://blog.collegevine.com/here-are-the-average-sat-scores-by-state/)
- [ACT](https://blog.prepscholar.com/act-scores-by-state-averages-highs-and-lows)

In [66]:
sat.describe()

Evidence-Based Reading and Writing        Math        Total
count                           51.000000   51.000000    51.000000
mean                           569.117647  547.627451  1126.098039
std                             45.666901   84.909119    92.494812
min                            482.000000   52.000000   950.000000
25%                            533.500000  522.000000  1055.500000
50%                            559.000000  548.000000  1107.000000
75%                            613.000000  599.000000  1212.000000
max                            644.000000  651.000000  1295.000000

In [67]:
act.describe(include=(np.object, np.number)) #not really sure what this means. 

State Participation    English       Math    Reading  \
count             52            52  52.000000  52.000000  52.000000   
unique            52            28        NaN        NaN        NaN   
top     North Dakota          100%        NaN        NaN        NaN   
freq               1            17        NaN        NaN        NaN   
mean             NaN           NaN  20.919231  21.173077  22.001923   
std              NaN           NaN   2.332132   1.963602   2.048672   
min              NaN           NaN  16.300000  18.000000  18.100000   
25%              NaN           NaN  19.000000  19.400000  20.475000   
50%              NaN           NaN  20.550000  20.900000  21.700000   
75%              NaN           NaN  23.300000  23.100000  24.125000   
max              NaN           NaN  25.500000  25.300000  26.000000   

          Science Composite  
count   52.000000        52  
unique        NaN        38  
top           NaN      20.3  
freq          NaN         3  
mean    21.040385       NaN  
std      3.151113       NaN  
min      2.300000       NaN  
25%     19.900000       NaN  
50%     21.150000       NaN  
75%     22.525000       NaN  
max     24.900000       NaN

Answer:

#### 4c. Fix any errors you identified

**The data is available** so there's no need to guess or calculate anything. If you didn't find any errors, continue to the next step.

In [68]:
sat.iloc[20,3] = 524 #I didnt run this till i mereged both dataframes

#### 5. What are your data types? 
Display the data types of each feature. 

In [69]:
#code
sat.dtypes

State                                 object
Participation                         object
Evidence-Based Reading and Writing     int64
Math                                   int64
Total                                  int64
dtype: object

In [70]:
act.dtypes

State             object
Participation     object
English          float64
Math             float64
Reading          float64
Science          float64
Composite         object
dtype: object

What did you learn?
- Do any of them seem odd?   
Sat has 5 columns. Act has 7 column
- Which ones are not as they should be?  
Participation is an object, should be an int 

Answer:

#### 6. Fix Incorrect Data Types
Based on what you discovered above, use appropriate methods to re-type incorrectly typed data.
- Define a function that will allow you to convert participation rates to an appropriate numeric type. Use `map` or `apply` to change these columns in each dataframe.

In [71]:
#code  change Participation in act to  
act.Participation = act.Participation.map(lambda x: float(x.replace('%',' ')))
act.tail()

State  Participation  English  Math  Reading  Science Composite
47       Virginia           29.0     23.5  23.3     24.6     23.5      23.8
48     Washington           29.0     20.9  21.9     22.1     22.0      21.9
49  West Virginia           69.0     20.0  19.4     21.2     20.5      20.4
50      Wisconsin          100.0     19.7  20.4     20.6     20.9      20.5
51        Wyoming          100.0     19.4  19.8     20.8     20.6     20.2x

In [72]:
def drop_percent(x):
    return float(x.replace('%',' ')) #def the function, name the function, inputs, return output(what do you want back)

In [73]:
drop_percent('48%')

48.0

In [74]:
# changed sat participation from object to float 
sat.Participation = sat.Participation.map(lambda x: float(x.replace('%',' ')))
sat.tail()

State  Participation  Evidence-Based Reading and Writing  Math  \
46       Virginia           65.0                                 561   541   
47     Washington           64.0                                 541   534   
48  West Virginia           14.0                                 558   528   
49      Wisconsin            3.0                                 642   649   
50        Wyoming            3.0                                 626   604   

    Total  
46   1102  
47   1075  
48   1086  
49   1291  
50   1230

In [75]:
act['Composite'] = act['Composite'].str.replace("x", "")
act.tail()

State  Participation  English  Math  Reading  Science Composite
47       Virginia           29.0     23.5  23.3     24.6     23.5      23.8
48     Washington           29.0     20.9  21.9     22.1     22.0      21.9
49  West Virginia           69.0     20.0  19.4     21.2     20.5      20.4
50      Wisconsin          100.0     19.7  20.4     20.6     20.9      20.5
51        Wyoming          100.0     19.4  19.8     20.8     20.6      20.2

In [76]:
#remove x from composite
act['Composite'] = act['Composite'].str.replace("x", "")
act.tail()

State  Participation  English  Math  Reading  Science Composite
47       Virginia           29.0     23.5  23.3     24.6     23.5      23.8
48     Washington           29.0     20.9  21.9     22.1     22.0      21.9
49  West Virginia           69.0     20.0  19.4     21.2     20.5      20.4
50      Wisconsin          100.0     19.7  20.4     20.6     20.9      20.5
51        Wyoming          100.0     19.4  19.8     20.8     20.6      20.2

In [77]:
#remove x from composite
act['Composite'] = act['Composite'].str.replace("x", "")
act.tail()

State  Participation  English  Math  Reading  Science Composite
47       Virginia           29.0     23.5  23.3     24.6     23.5      23.8
48     Washington           29.0     20.9  21.9     22.1     22.0      21.9
49  West Virginia           69.0     20.0  19.4     21.2     20.5      20.4
50      Wisconsin          100.0     19.7  20.4     20.6     20.9      20.5
51        Wyoming          100.0     19.4  19.8     20.8     20.6      20.2

In [78]:
act['Composite'] = act['Composite'].str.replace("x", "")
act.tail()

State  Participation  English  Math  Reading  Science Composite
47       Virginia           29.0     23.5  23.3     24.6     23.5      23.8
48     Washington           29.0     20.9  21.9     22.1     22.0      21.9
49  West Virginia           69.0     20.0  19.4     21.2     20.5      20.4
50      Wisconsin          100.0     19.7  20.4     20.6     20.9      20.5
51        Wyoming          100.0     19.4  19.8     20.8     20.6      20.2

In [79]:
act['Composite'] = act.Composite.astype(float)

In [80]:
#just looking at some stats
#act.groupby('Participation', axis=0).mean()

In [81]:
type(act['Composite'])

pandas.core.series.Series

In [82]:
type(act.iloc[0,1])

numpy.float64

- Fix any individual values preventing other columns from being the appropriate type.

- Finish your data modifications by making sure the columns are now typed appropriately.

In [83]:
#code
sat.dtypes

State                                  object
Participation                         float64
Evidence-Based Reading and Writing      int64
Math                                    int64
Total                                   int64
dtype: object

- Display the data types again to confirm they are correct.

In [84]:
act.dtypes

State             object
Participation    float64
English          float64
Math             float64
Reading          float64
Science          float64
Composite        float64
dtype: object

In [85]:
#sat['sat_participation'] = sat.sat_participation.astype(float)

#### 7. Rename Columns
Change the names of the columns to more expressive names so that you can tell the difference the SAT columns and the ACT columns. Your solution should map all column names being changed at once (no repeated singular name-changes). **We will be combining these data with some of the data from 2018, and so you should name columns in an appropriate way**.

**Guidelines**:
- Column names should be all lowercase (you will thank yourself when you start pushing data to SQL later in the course)
- Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`.
- Column names should be unique and informative (the only feature that we actually share between dataframes is the state).

In [86]:
sat_new_columns = {
    "State": "state",
    "Participation": "sat_participation_2017",
    "Evidence-Based Reading and Writing": "sat_reading_2017",
    "Math": "sat_math_2017",
    "Total": "sat_total_2017"
}

sat.rename(columns=sat_new_columns, inplace=True)

In [87]:
sat.columns

Index(['state', 'sat_participation_2017', 'sat_reading_2017', 'sat_math_2017',
       'sat_total_2017'],
      dtype='object')

In [88]:
act_new_columns = {
    "State": "state",
    "Participation": "act_participation_2017",
    "English": "act_english_2017",
    "Math": "act_math_2017",
    "Reading": "act_reading_2017",
    "Science": "act_science_2017",
    "Composite": "act_composite_2017"
}

act.rename(columns=act_new_columns, inplace=True)

In [89]:
act.columns

Index(['state', 'act_participation_2017', 'act_english_2017', 'act_math_2017',
       'act_reading_2017', 'act_science_2017', 'act_composite_2017'],
      dtype='object')

#### 8. Create a data dictionary

Now that we've fixed our data, and given it appropriate names, let's create a [data dictionary](http://library.ucmerced.edu/node/10249). 

A data dictionary provides a quick overview of features/variables/columns, alongside data types and descriptions. The more descriptive you can be, the more useful this document is.

Example of a Fictional Data Dictionary Entry: 

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**county_pop**|*integer*|2010 census|The population of the county (units in thousands, where 2.5 represents 2500 people).| 
|**per_poverty**|*float*|2010 census|The percent of the county over the age of 18 living below the 200% of official US poverty rate (units percent to two decimal places 98.10 means 98.1%)|

[Here's a quick link to a short guide for formatting markdown in Jupyter notebooks](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html).

Provided is the skeleton for formatting a markdown table, with columns headers that will help you create a data dictionary to quickly summarize your data, as well as some examples. **This would be a great thing to copy and paste into your custom README for this project.**

|Feature|Type|Dataset|Description|
|---|---|---|---|
|column name|int/float/object|ACT/SAT|This is an example| 


|Feature|Type|Dataset|Description|
|---|---|---|---|
|**State**|*string*|SAT/ACT|State| 
|**SAT Participation 2017**|*int/float/object*|SAT/ACT| Paricipation rate for each state|
|**SAT Reading 2017**|*int/float/object*|SAT/ACT| Reading scores for each state|
|**SAT Math 2017**|*int/float/object*|SAT/ACT| Math scores for each state
|**SAT Total 2017**|*int/float/object*|SAT/ACT| SAT total scores for each state
|**ACT Participation 2017**|*int/float/object*|SAT/ACT| Paricipation rate for each state|
|**ACT English 2017**|*int/float/object*|SAT/ACT| English score for each state
|**ACT Math 2017**|*int/float/object*|SAT/ACT| Math score for each state
|**ACT Reading 2017**|*int/float/object*|SAT/ACT| Reading score for each state
|**ACT Science 2017**|*int/float/object*|SAT/ACT| Science score for each state
|**ACT Composite 2017**|*int/float/object*|SAT/ACT| ACT Composite for each state

In [90]:
act.head(1)

state  act_participation_2017  act_english_2017  act_math_2017  \
0  National                    60.0              20.3           20.7   

   act_reading_2017  act_science_2017  act_composite_2017  
0              21.4              21.0                21.0

#### 9. Drop unnecessary rows

One of our dataframes contains an extra row. Identify and remove this from the dataframe.

In [91]:
#code
#act.shape
act.drop(0,inplace=True)

In [92]:
sat.shape

(51, 5)

In [93]:
act.shape

(51, 7)

#### 10. Merge Dataframes

Join the 2017 ACT and SAT dataframes using the state in each dataframe as the key. Assign this to a new variable.

In [94]:
#Code:
#couldnt get this to work.
# frames = [df1, df2]
# result = pd.concat(frames, ignore_index=True)
# frames= [sat,act]
# df_2017 = pd.concat(frames, ignore_index=False, sort=False)

In [95]:

df_2017 = pd.merge(sat, act, how='left', on='state', suffixes=('SAT','ACT'))
df_2017.head(21)

state  sat_participation_2017  sat_reading_2017  \
0                Alabama                     5.0               593   
1                 Alaska                    38.0               547   
2                Arizona                    30.0               563   
3               Arkansas                     3.0               614   
4             California                    53.0               531   
5               Colorado                    11.0               606   
6            Connecticut                   100.0               530   
7               Delaware                   100.0               503   
8   District of Columbia                   100.0               482   
9                Florida                    83.0               520   
10               Georgia                    61.0               535   
11                Hawaii                    55.0               544   
12                 Idaho                    93.0               513   
13              Illinois                     9.0               559   
14               Indiana                    63.0               542   
15                  Iowa                     2.0               641   
16                Kansas                     4.0               632   
17              Kentucky                     4.0               631   
18             Louisiana                     4.0               611   
19                 Maine                    95.0               513   
20              Maryland                    69.0               536   

    sat_math_2017  sat_total_2017  act_participation_2017  act_english_2017  \
0             572            1165                   100.0              18.9   
1             533            1080                    65.0              18.7   
2             553            1116                    62.0              18.6   
3             594            1208                   100.0              18.9   
4             524            1055                    31.0              22.5   
5             595            1201                   100.0              20.1   
6             512            1041                    31.0              25.5   
7             492             996                    18.0              24.1   
8             468             950                    32.0              24.4   
9             497            1017                    73.0              19.0   
10            515            1050                    55.0              21.0   
11            541            1085                    90.0              17.8   
12            493            1005                    38.0              21.9   
13            556            1115                    93.0              21.0   
14            532            1074                    35.0              22.0   
15            635            1275                    67.0              21.2   
16            628            1260                    73.0              21.1   
17            616            1247                   100.0              19.6   
18            586            1198                   100.0              19.4   
19            499            1012                     8.0              24.2   
20            524            1060                    28.0              23.3   

    act_math_2017  act_reading_2017  act_science_2017  act_composite_2017  
0            18.4              19.7              19.4                19.2  
1            19.8              20.4              19.9                19.8  
2            19.8              20.1              19.8                19.7  
3            19.0              19.7              19.5                19.4  
4            22.7              23.1              22.2                22.8  
5            20.3              21.2              20.9                20.8  
6            24.6              25.6              24.6                25.2  
7            23.4              24.8              23.6                24.1  
8            23.5              24.9              23.5  

In [96]:
df_2017.head(1)

state  sat_participation_2017  sat_reading_2017  sat_math_2017  \
0  Alabama                     5.0               593            572   

   sat_total_2017  act_participation_2017  act_english_2017  act_math_2017  \
0            1165                   100.0              18.9           18.4   

   act_reading_2017  act_science_2017  act_composite_2017  
0              19.7              19.4                19.2

In [97]:
# just thinking to hard about this 52
# df_2017 = df_2017.sat_math.map(lambda x: 520 if x == 52 else int(x))
# df_2017.iloc[20,3] This fucked up all my data 2017 

In [98]:
df_2017_agg = df_2017.agg({
    'sat_participation_2017': ['min','max'],
    'sat_reading_2017': ['min','max'],
    'sat_math_2017': ['min','max'],
    'sat_total_2017': ['min','max'],
    'act_participation_2017': ['min','max'],
    'act_english_2017': ['min','max'],
    'act_math_2017': ['min','max'],
    'act_reading_2017': ['min','max'],
    'act_science_2017': ['min','max'],
    'act_composite_2017': ['min','max'],
})
df_2017_agg

sat_participation_2017  sat_reading_2017  sat_math_2017  sat_total_2017  \
min                     2.0               482            468             950   
max                   100.0               644            651            1295   

     act_participation_2017  act_english_2017  act_math_2017  \
min                     8.0              16.3           18.0   
max                   100.0              25.5           25.3   

     act_reading_2017  act_science_2017  act_composite_2017  
min              18.1               2.3                17.8  
max              26.0              24.9                25.5

#### 11. Save your cleaned, merged dataframe

Use a relative path to save out your data as `combined_2017.csv`.

In [99]:
df_2017.to_csv('combined_2017.csv', header=True, index=True)

In [100]:
df_2017.head()

state  sat_participation_2017  sat_reading_2017  sat_math_2017  \
0     Alabama                     5.0               593            572   
1      Alaska                    38.0               547            533   
2     Arizona                    30.0               563            553   
3    Arkansas                     3.0               614            594   
4  California                    53.0               531            524   

   sat_total_2017  act_participation_2017  act_english_2017  act_math_2017  \
0            1165                   100.0              18.9           18.4   
1            1080                    65.0              18.7           19.8   
2            1116                    62.0              18.6           19.8   
3            1208                   100.0              18.9           19.0   
4            1055                    31.0              22.5           22.7   

   act_reading_2017  act_science_2017  act_composite_2017  
0              19.7              19.4                19.2  
1              20.4              19.9                19.8  
2              20.1              19.8                19.7  
3              19.7              19.5                19.4  
4              23.1              22.2                22.8

## 2018 Data Import and Cleaning

Links to the 2018 ACT and SAT data are provided in the README. These data live in PDFs, and so you'll get to enjoy practicing some *manual* data collection. Save these data as a CSV in your `data` directory, and import, explore, and clean these data in the same way you did above. **Make sure you comment on your steps so it is clear *why* you are doing each process**.

#### Combine your 2017 and 2018 data into a single dataframe
Joining on state names should work, assuming you formatted all your state names identically. Make sure none of your columns (other than state) have identical names. Do yourself a favor and decide if you're encoding participation rates as floats or integers and standardize this across your datasets.

Save the contents of this merged dataframe as `final.csv`.

**Use this combined dataframe for the remainder of the project**.

In [101]:
sat_2018 = pd.read_csv(r'/Users/owner/Desktop/gclone/project_1/project_1/project-1/data/East_Coast_Data_Entry - SAT 2018.csv')
sat_2018.head()

State Participation  Evidence-Based Reading and Writing  Math  Total
0     Alabama            6%                                 595   571   1166
1      Alaska           43%                                 562   544   1106
2     Arizona           29%                                 577   572   1149
3    Arkansas            5%                                 592   576   1168
4  California           60%                                 540   536   1076

In [102]:
act_2018 = pd.read_csv(r'/Users/owner/Desktop/gclone/project_1/project_1/project-1/data/East_Coast_Data_Entry - ACT 2018.csv')
act_2018.head()

State Participation  Composite
0     Alabama          100%       19.1
1      Alaska           33%       20.8
2     Arizona           66%       19.2
3    Arkansas          100%       19.4
4  California           27%       22.7

In [103]:
act_2018.shape 

(52, 3)

In [104]:
sat_2018.shape

(52, 5)

In [105]:
sat_2018.head(10)

State Participation  Evidence-Based Reading and Writing  \
0               Alabama            6%                                 595   
1                Alaska           43%                                 562   
2               Arizona           29%                                 577   
3              Arkansas            5%                                 592   
4            California           60%                                 540   
5              Colorado           28%                                 519   
6           Connecticut          100%                                 535   
7              Delaware          100%                                 505   
8  District of Columbia           92%                                 497   
9               Florida           97%                                 522   

   Math  Total  
0   571   1166  
1   544   1106  
2   572   1149  
3   576   1168  
4   536   1076  
5   506   1025  
6   519   1054  
7   492    997  
8   480    977  
9   493   1015

In [106]:
act_2018.head(10)

State Participation  Composite
0           Alabama          100%       19.1
1            Alaska           33%       20.8
2           Arizona           66%       19.2
3          Arkansas          100%       19.4
4        California           27%       22.7
5          Colorado           30%       23.9
6       Connecticut           26%       25.6
7          Delaware           17%       23.8
8  Washington, D.C.           32%       23.6
9           Florida           66%       19.9

In [107]:
#thx to everyone on slack they brought this to my attention
act_2018.iloc[8, 0] = act_2018.iloc[8, 0].replace('Washington, D.C.', 'District of Columbia')

In [108]:
act_2018.dtypes

State             object
Participation     object
Composite        float64
dtype: object

#### Answer:
sat_2018: 
    -I see no NaN values. (.isna)
    -It is a DataFrame (type())
    -No duplicate values (.duplicate)
    -Wyoming has the max values in participation, reading-writing, and math (.max)
    -Alabama has the lowest scores, but not the lowest participation rate. (.min) 
    -shape of 51 rows and 5 columns (.shape) 
    -Missing total Averages
    -sat_2018 gives 
act_2018: 
    -I see no NaN values. (.isna)
    -It is a DataFrame (type())
    -No duplicate values (.duplicate)
    -Wyoming has the max values in participation, reading-writing, and math (.max)
    -Alabama has the lowest scores, but not the lowest participation rate. Alabama's participation is much higherhere      than SAT participation  (.min) 
    - ACT gives us national AVG. 

In [109]:
sat_2018.head()

State Participation  Evidence-Based Reading and Writing  Math  Total
0     Alabama            6%                                 595   571   1166
1      Alaska           43%                                 562   544   1106
2     Arizona           29%                                 577   572   1149
3    Arkansas            5%                                 592   576   1168
4  California           60%                                 540   536   1076

In [110]:
#code  change Participation in sat_2018 to a float with out the '%'  
sat_2018.Participation = sat_2018.Participation.map(lambda x: float(x.replace('%',' ')))
sat_2018.tail()

State  Participation  Evidence-Based Reading and Writing  Math  \
47     Washington           69.0                                 543   538   
48  West Virginia           28.0                                 513   486   
49      Wisconsin            3.0                                 641   653   
50        Wyoming            3.0                                 633   625   
51       National           51.0                                 568   558   

    Total  
47   1081  
48    999  
49   1294  
50   1258  
51   1126

In [111]:
#code  change Participation in sat_2018 to a float with out the '%'  
act_2018.Participation = act_2018.Participation.map(lambda x: float(x.replace('%',' ')))
act_2018.tail()

State  Participation  Composite
47     Washington           24.0       22.2
48  West Virginia           65.0       20.3
49      Wisconsin          100.0       20.5
50        Wyoming          100.0       20.0
51       National           50.0       21.5

In [112]:
sat_2018.tail()

State  Participation  Evidence-Based Reading and Writing  Math  \
47     Washington           69.0                                 543   538   
48  West Virginia           28.0                                 513   486   
49      Wisconsin            3.0                                 641   653   
50        Wyoming            3.0                                 633   625   
51       National           51.0                                 568   558   

    Total  
47   1081  
48    999  
49   1294  
50   1258  
51   1126

#### 5. What are your data types? 
Display the data types of each feature. 

In [113]:
df_2017.dtypes

state                      object
sat_participation_2017    float64
sat_reading_2017            int64
sat_math_2017               int64
sat_total_2017              int64
act_participation_2017    float64
act_english_2017          float64
act_math_2017             float64
act_reading_2017          float64
act_science_2017          float64
act_composite_2017        float64
dtype: object

In [114]:
sat_2018.dtypes

State                                  object
Participation                         float64
Evidence-Based Reading and Writing      int64
Math                                    int64
Total                                   int64
dtype: object

In [115]:
act_2018.dtypes

State             object
Participation    float64
Composite        float64
dtype: object

**What is the minimum *possible* value for each test/subtest? What is the maximum *possible* value?**


In [116]:
sat_2018.describe()

Participation  Evidence-Based Reading and Writing        Math  \
count      52.000000                           52.000000   52.000000   
mean       45.346154                          567.711538  558.442308   
std        36.941865                           44.864649   48.977551   
min         2.000000                          497.000000  480.000000   
25%         4.750000                          535.000000  521.750000   
50%        49.500000                          558.000000  547.000000   
75%        76.750000                          615.750000  598.000000   
max       100.000000                          643.000000  655.000000   

             Total  
count    52.000000  
mean   1126.153846  
std      93.513632  
min     977.000000  
25%    1063.500000  
50%    1102.500000  
75%    1215.000000  
max    1298.000000

In [117]:
sat_2018.describe(include=(np.object, np.number, np.float))

State  Participation  Evidence-Based Reading and Writing  \
count             52      52.000000                           52.000000   
unique            52            NaN                                 NaN   
top     North Dakota            NaN                                 NaN   
freq               1            NaN                                 NaN   
mean             NaN      45.346154                          567.711538   
std              NaN      36.941865                           44.864649   
min              NaN       2.000000                          497.000000   
25%              NaN       4.750000                          535.000000   
50%              NaN      49.500000                          558.000000   
75%              NaN      76.750000                          615.750000   
max              NaN     100.000000                          643.000000   

              Math        Total  
count    52.000000    52.000000  
unique         NaN          NaN  
top            NaN          NaN  
freq           NaN          NaN  
mean    558.442308  1126.153846  
std      48.977551    93.513632  
min     480.000000   977.000000  
25%     521.750000  1063.500000  
50%     547.000000  1102.500000  
75%     598.000000  1215.000000  
max     655.000000  1298.000000

In [118]:
act_2018.describe(include=(np.object, np.number, np.float))

State  Participation  Composite
count             52      52.000000  52.000000
unique            52            NaN        NaN
top     North Dakota            NaN        NaN
freq               1            NaN        NaN
mean             NaN      61.500000  21.496154
std              NaN      33.740939   2.090779
min              NaN       7.000000  17.700000
25%              NaN      29.250000  19.975000
50%              NaN      65.500000  21.300000
75%              NaN     100.000000  23.625000
max              NaN     100.000000  25.600000

#### Rename the columns

In [119]:
df_2017.head()

state  sat_participation_2017  sat_reading_2017  sat_math_2017  \
0     Alabama                     5.0               593            572   
1      Alaska                    38.0               547            533   
2     Arizona                    30.0               563            553   
3    Arkansas                     3.0               614            594   
4  California                    53.0               531            524   

   sat_total_2017  act_participation_2017  act_english_2017  act_math_2017  \
0            1165                   100.0              18.9           18.4   
1            1080                    65.0              18.7           19.8   
2            1116                    62.0              18.6           19.8   
3            1208                   100.0              18.9           19.0   
4            1055                    31.0              22.5           22.7   

   act_reading_2017  act_science_2017  act_composite_2017  
0              19.7              19.4                19.2  
1              20.4              19.9                19.8  
2              20.1              19.8                19.7  
3              19.7              19.5                19.4  
4              23.1              22.2                22.8

In [120]:
sat_2018.head()

State  Participation  Evidence-Based Reading and Writing  Math  Total
0     Alabama            6.0                                 595   571   1166
1      Alaska           43.0                                 562   544   1106
2     Arizona           29.0                                 577   572   1149
3    Arkansas            5.0                                 592   576   1168
4  California           60.0                                 540   536   1076

In [121]:
sat2018_new_columns = {
    "State": "state",
    "Participation": "sat_participation_2018",
    "Evidence-Based Reading and Writing": "sat_reading_2018",
    "Math": "sat_math_2018",
    "Total": "sat_total_2018"
}

sat_2018.rename(columns=sat2018_new_columns, inplace=True)

In [122]:
act2018_new_columns = {
    "State": "state",
    "Participation": "act_participation_2018",
    "English": "act_english_2018",
    "Math": "act_math_2018",
    "Reading": "act_reading_2018",
    "Science": "act_science_2018",
    "Composite": "act_composite_2018"
}

act_2018.rename(columns=act2018_new_columns, inplace=True)

In [123]:
act_2018.columns

Index(['state', 'act_participation_2018', 'act_composite_2018'], dtype='object')

#### 9. Drop unnecessary rows

In [124]:
sat_2018.tail()

state  sat_participation_2018  sat_reading_2018  sat_math_2018  \
47     Washington                    69.0               543            538   
48  West Virginia                    28.0               513            486   
49      Wisconsin                     3.0               641            653   
50        Wyoming                     3.0               633            625   
51       National                    51.0               568            558   

    sat_total_2018  
47            1081  
48             999  
49            1294  
50            1258  
51            1126

In [125]:
sat_2018.drop(51, inplace=True)
sat_2018.tail()

state  sat_participation_2018  sat_reading_2018  sat_math_2018  \
46       Virginia                    68.0               567            550   
47     Washington                    69.0               543            538   
48  West Virginia                    28.0               513            486   
49      Wisconsin                     3.0               641            653   
50        Wyoming                     3.0               633            625   

    sat_total_2018  
46            1117  
47            1081  
48             999  
49            1294  
50            1258

In [126]:
act_2018.drop(51, inplace=True)
act_2018.tail()

state  act_participation_2018  act_composite_2018
46       Virginia                    24.0                23.9
47     Washington                    24.0                22.2
48  West Virginia                    65.0                20.3
49      Wisconsin                   100.0                20.5
50        Wyoming                   100.0                20.0

In [127]:
df_2017.tail()

state  sat_participation_2017  sat_reading_2017  sat_math_2017  \
46       Virginia                    65.0               561            541   
47     Washington                    64.0               541            534   
48  West Virginia                    14.0               558            528   
49      Wisconsin                     3.0               642            649   
50        Wyoming                     3.0               626            604   

    sat_total_2017  act_participation_2017  act_english_2017  act_math_2017  \
46            1102                    29.0              23.5           23.3   
47            1075                    29.0              20.9           21.9   
48            1086                    69.0              20.0           19.4   
49            1291                   100.0              19.7           20.4   
50            1230                   100.0              19.4           19.8   

    act_reading_2017  act_science_2017  act_composite_2017  
46              24.6              23.5                23.8  
47              22.1              22.0                21.9  
48              21.2              20.5                20.4  
49              20.6              20.9                20.5  
50              20.8              20.6                20.2

### Merge the data

In [128]:
act_2018.head() # just checking, gotta remind myself

state  act_participation_2018  act_composite_2018
0     Alabama                   100.0                19.1
1      Alaska                    33.0                20.8
2     Arizona                    66.0                19.2
3    Arkansas                   100.0                19.4
4  California                    27.0                22.7

In [129]:
sat_2018.head()

state  sat_participation_2018  sat_reading_2018  sat_math_2018  \
0     Alabama                     6.0               595            571   
1      Alaska                    43.0               562            544   
2     Arizona                    29.0               577            572   
3    Arkansas                     5.0               592            576   
4  California                    60.0               540            536   

   sat_total_2018  
0            1166  
1            1106  
2            1149  
3            1168  
4            1076

In [130]:
df_2018 = pd.merge(sat_2018, act_2018, how='left', on='state')
#df_2018.iloc[10:15,:]

In [131]:
df_2017.iloc[10:15,:]

state  sat_participation_2017  sat_reading_2017  sat_math_2017  \
10   Georgia                    61.0               535            515   
11    Hawaii                    55.0               544            541   
12     Idaho                    93.0               513            493   
13  Illinois                     9.0               559            556   
14   Indiana                    63.0               542            532   

    sat_total_2017  act_participation_2017  act_english_2017  act_math_2017  \
10            1050                    55.0              21.0           20.9   
11            1085                    90.0              17.8           19.2   
12            1005                    38.0              21.9           21.8   
13            1115                    93.0              21.0           21.2   
14            1074                    35.0              22.0           22.4   

    act_reading_2017  act_science_2017  act_composite_2017  
10              22.0              21.3                21.4  
11              19.2              19.3                19.0  
12              23.0              22.1                22.3  
13              21.6              21.3                21.4  
14              23.2              22.3                22.6

In [132]:
df_2017_2018=pd.merge(df_2017,df_2018, on='state')
df_2017_2018.head()

state  sat_participation_2017  sat_reading_2017  sat_math_2017  \
0     Alabama                     5.0               593            572   
1      Alaska                    38.0               547            533   
2     Arizona                    30.0               563            553   
3    Arkansas                     3.0               614            594   
4  California                    53.0               531            524   

   sat_total_2017  act_participation_2017  act_english_2017  act_math_2017  \
0            1165                   100.0              18.9           18.4   
1            1080                    65.0              18.7           19.8   
2            1116                    62.0              18.6           19.8   
3            1208                   100.0              18.9           19.0   
4            1055                    31.0              22.5           22.7   

   act_reading_2017  act_science_2017  act_composite_2017  \
0              19.7              19.4                19.2   
1              20.4              19.9                19.8   
2              20.1              19.8                19.7   
3              19.7              19.5                19.4   
4              23.1              22.2                22.8   

   sat_participation_2018  sat_reading_2018  sat_math_2018  sat_total_2018  \
0                     6.0               595            571            1166   
1                    43.0               562            544            1106   
2                    29.0               577            572            1149   
3                     5.0               592            576            1168   
4                    60.0               540            536            1076   

   act_participation_2018  act_composite_2018  
0                   100.0                19.1  
1                    33.0                20.8  
2                    66.0                19.2  
3                   100.0                19.4  
4                    27.0                22.7

In [133]:
df_2017_2018.dtypes

state                      object
sat_participation_2017    float64
sat_reading_2017            int64
sat_math_2017               int64
sat_total_2017              int64
act_participation_2017    float64
act_english_2017          float64
act_math_2017             float64
act_reading_2017          float64
act_science_2017          float64
act_composite_2017        float64
sat_participation_2018    float64
sat_reading_2018            int64
sat_math_2018               int64
sat_total_2018              int64
act_participation_2018    float64
act_composite_2018        float64
dtype: object

## Exploratory Data Analysis


### Summary Statistics
Transpose the output of pandas `describe` method to create a quick overview of each numeric feature.

In [134]:
#Code:
df_2017_2018.describe().T


count         mean        std    min      25%     50%  \
sat_participation_2017   51.0    39.803922  35.276632    2.0     4.00    38.0   
sat_reading_2017         51.0   569.117647  45.666901  482.0   533.50   559.0   
sat_math_2017            51.0   556.882353  47.121395  468.0   523.50   548.0   
sat_total_2017           51.0  1126.098039  92.494812  950.0  1055.50  1107.0   
act_participation_2017   51.0    65.254902  32.140842    8.0    31.00    69.0   
act_english_2017         51.0    20.931373   2.353677   16.3    19.00    20.7   
act_math_2017            51.0    21.182353   1.981989   18.0    19.40    20.9   
act_reading_2017         51.0    22.013725   2.067271   18.1    20.45    21.8   
act_science_2017         51.0    21.041176   3.182463    2.3    19.90    21.3   
act_composite_2017       51.0    21.519608   2.020695   17.8    19.80    21.4   
sat_participation_2018   51.0    45.235294  37.300718    2.0     4.50    48.0   
sat_reading_2018         51.0   567.705882  45.311056  497.0   535.00   554.0   
sat_math_2018            51.0   558.450980  49.464862  480.0   521.50   547.0   
sat_total_2018           51.0  1126.156863  94.444136  977.0  1063.00  1099.0   
act_participation_2018   51.0    61.725490  34.037085    7.0    28.50    66.0   
act_composite_2018       51.0    21.496078   2.111583   17.7    19.95    21.3   

                            75%     max  
sat_participation_2017    66.00   100.0  
sat_reading_2017         613.00   644.0  
sat_math_2017            599.00   651.0  
sat_total_2017          1212.00  1295.0  
act_participation_2017   100.00   100.0  
act_english_2017          23.30    25.5  
act_math_2017             23.10    25.3  
act_reading_2017          24.15    26.0  
act_science_2017          22.75    24.9  
act_composite_2017        23.60    25.5  
sat_participation_2018    77.50   100.0  
sat_reading_2018         616.50   643.0  
sat_math_2018            601.00   655.0  
sat_total_2018          1220.00  1298.0  
act_participation_2018   100.00   100.0  
act_composite_2018        23.65    25.6

#### Manually calculate standard deviation

$$\sigma = \sqrt{\frac{1}{n}\sum_{i=1}^n(x_i - \mu)^2}$$

- Write a function to calculate standard deviation using the formula above

In [135]:
# import math
# def mean(data):
#     return float(sum(data) / len(data))

# def variance(data):
#     mu = mean(data)
#     return mean([(x - mu) ** 2 for x in data])

# def stddev(data,column_name):
#     return math.sqrt(variance(data))
# stddev(df_2017_2018,'sat_reading_2017')

In [136]:
# def std(dataframe, column_name):
#     mu = dataframe[column_name].mean()
#     var = sum([((x-mu)**2) for x in dataframe[column_name]])/len(dataframe)
#     std = var**0.5
#     return std
# std(df_2017_2018)   #joey told me we didnt need the column_name and made it work 

In [140]:
df_2017_2018.dtypes

state                      object
sat_participation_2017    float64
sat_reading_2017            int64
sat_math_2017               int64
sat_total_2017              int64
act_participation_2017    float64
act_english_2017          float64
act_math_2017             float64
act_reading_2017          float64
act_science_2017          float64
act_composite_2017        float64
sat_participation_2018    float64
sat_reading_2018            int64
sat_math_2018               int64
sat_total_2018              int64
act_participation_2018    float64
act_composite_2018        float64
dtype: object

In [141]:
def std(dataframe):
    mu = dataframe.mean()
    var = sum([((x-mu)**2) for x in dataframe])/len(dataframe)
    std = var**0.5
    return std
std(df_2017_2018['sat_math_2018'])

48.97751133618255

In [ ]:
df

In [ ]:
df_2017_2018_agg= df_2017_2018.agg("mean", axis="index")
df_2017_2018_agg

In [ ]:
# varience_all = df_2017_2018_agg.var(axis=0)

# varience_all.

In [ ]:
# no_state = df_2017_2018.drop(['state'], axis=1)
# no_state.head()

In [ ]:
# import math
# def stddev(data):
#     variance=varience_all
#     return math.sqrt(variance(data))

In [ ]:
# type(varience_all)

In [ ]:
# for i in varience_all:
#     print(math.sqrt(i)) 

- Use a **dictionary comprehension** to apply your standard deviation function to each numeric column in the dataframe.  **No loops**  
- Assign the output to variable `sd` as a dictionary where: 
    - Each column name is now a key 
    - That standard deviation of the column is the value 
     
*Example Output :* `{'ACT_Math': 120, 'ACT_Reading': 120, ...}`

In [ ]:
df_2017_2018.head()

In [ ]:
mask = (
    {'sat_participation_2017': std(df_2017_2018['sat_participation_2017'])},
    {'sat_reading_2017' : std(df_2017_2018['sat_reading_2017'])},
    {'sat_math_2017' : std(df_2017_2018['sat_math_2017'])},
    {'' : std(df_2017_2018[''])},
)
mask  I stopped here and searched for a way i didnt have to type all this shit out 

In [ ]:
sd= df_2017_2018.iloc[:,1:]
sd.head()

In [ ]:
sd_2= sd.apply(std)
sd_2_dict = dict(sd_2)

In [ ]:
sd_2_dict

Do your manually calculated standard deviations match up with the output from pandas `describe`? What about numpy's `std` method?

In [ ]:
np.mean(df_2017_2018) #

Answer: 

#### Investigate trends in the data
Using sorting and/or masking (along with the `.head` method to not print our entire dataframe), consider the following questions:

- Which states have the highest and lowest participation rates for the:
    - 2017 SAT?
    - 2018 SAT?
    - 2017 ACT?
    - 2018 ACT?
- Which states have the highest and lowest mean total/composite scores for the:
    - 2017 SAT?
    - 2018 SAT?
    - 2017 ACT?
    - 2018 ACT?
- Do any states with 100% participation on a given test have a rate change year-to-year?
- Do any states show have >50% participation on *both* tests either year?

Based on what you've just observed, have you identified any states that you're especially interested in? **Make a note of these and state *why* you think they're interesting**.

**You should comment on your findings at each step in a markdown cell below your code block**. Make sure you include at least one example of sorting your dataframe by a column, and one example of using boolean filtering (i.e., masking) to select a subset of the dataframe.

In [ ]:
# get minimum and maximum participation rate for SAT and ACT for 2017:

min(df_2017_2018['sat_participation_2017'])

In [ ]:
# 2017 SAT participation max,min depending on the ascending
df_2017_2018.sort_values(by='sat_participation_2017', ascending=True).head(10)#[['state', 'sat_participation_2017']]

In [ ]:
# 2018 SAT participation max, min depending on the ascneding
df_2017_2018.sort_values(by='sat_participation_2018', ascending=True).head(10)

In [ ]:
#2017 ACT parti max,min depending on the ascending 
df_2017_2018.sort_values(by='act_participation_2017', ascending=True).head(10)

In [ ]:
# 2018 ACT parti max,min depending on the ascending
df_2017_2018.sort_values(by='act_participation_2018', ascending=True).head(10)

In [ ]:
#Which states have the highest and lowest mean total/composite scores for the:
df_2017_2018.sort_values(by='sat_total_2017', ascending=False).tail(10)

In [ ]:
#another fuck up 
#(df_2017_2018.groupby('sat_participation_2017').min().head()) & (df_2017_2018.groupby('act_participation_2017').min().head())

In [ ]:
#max,min for 2018(change the asceding to get the min)
df_2017_2018.sort_values(by='act_composite_2018', ascending=False).head(3)

In [ ]:
##max,min for 2017 (change the asceding to get the min)
df_2017_2018.sort_values(by='act_composite_2017', ascending=False).head(3)

In [ ]:
#max,min for 2017. (change the asceding to get the min)
df_2017_2018.sort_values(by='sat_total_2017', ascending=False).head(3)

In [ ]:
#max,min for 2018(change the asceding to get the min)
df_2017_2018.sort_values(by='sat_total_2018', ascending=False).head(3)

In [ ]:
act_2018.dtypes

In [ ]:
#Which states have the highest and lowest mean total/composite scores for the:
# this gives me that total, not the states
df_2017_2018_mean = df_2017_2018.agg({
    'sat_total_2017': ['mean'],
    'act_composite_2018': ['mean'],
})
df_2017_2018_mean

#df_2017_2018.loc[:,'sat_total_2017'].mean() another way of saying the same thing, but for a single column

In [ ]:
df_2017_2018[['sat_total_2017']].mean(axis=1)
test = df_2017_2018.sort_values(by='sat_total_2017', ascending=True).mean()
test.head() # not even sure what this does 

In [ ]:
df_2017_2018.loc[:,'sat_total_2017'].mean()

In [ ]:
df_2017_2018_agg = df_2017_2018.agg({
    'sat_participation_2017': ['min','max','mean','std','median'],
    'sat_reading_2017': ['min','max','mean','std','median'],
    'sat_math_2017': ['min','max','mean','std','median'],
    'sat_total_2017': ['min','max','mean','std','median'],
    'act_participation_2017': ['min','max','mean','std','median'],
    'act_english_2017': ['min','max','mean','std','median'],
    'act_math_2017': ['min','max','mean','std','median'],
    'act_reading_2017': ['min','max','mean','std','median'],
    'act_science_2017': ['min','max','mean','std','median'],
    'act_composite_2017': ['min','max','mean','std','median'],
    'sat_participation_2018': ['min','max','mean','std','median'],
    'sat_reading_2018': ['min','max','mean','std','median'],
    'sat_math_2018': ['min','max','mean','std','median'],
    'sat_total_2018': ['min','max','mean','std','median'],
    'act_participation_2018': ['min','max','mean','std','median'],
    'act_composite_2018': ['min','max','mean','std','median'],
})
df_2017_2018_agg.T

In [ ]:
df_2018_agg = df_2017_2018.agg({
    'sat_participation_2018': ['min','max','mean','median','std'],
    'sat_reading_2018': ['min','max','mean','median'],
    'sat_math_2018': ['min','max','mean','median'],
    'sat_total_2018': ['min','max','mean','median'],
    'act_participation_2018': ['min','max','mean','median'],
    'act_composite_2018': ['min','max','mean','median'],
})
df_2018_agg.T

In [ ]:
#what states show >50% participation on both tests for 2017?
participation_total_2017 = df_2017_2018[(df_2017_2018['sat_participation_2017'] > 60) & (df_2017_2018['act_participation_2017'] > 50)][0:50]
participation_total_2017

In [ ]:
#satparticipation rate of 50 or more and score 1100 or greater 
state_part_comp = df_2017_2018[(df_2017_2018['sat_participation_2017'] > 50) & (df_2017_2018['sat_total_2017'] > 1100)][0:50]
state_part_comp

In [ ]:
#Do any states with 100% participation on a given test have a rate change year-to-year?
#D.C had an 8% drop YoY
#Florida had the highest sat_participation absolute change YoY among all the states that had 80%+
#Colorado offers free SAT but didnt even get an 80% participation rate. In 2017 thier participation rate was 11% in 2017 then increased to 28%

participation_sat_2017 = df_2017_2018[(df_2017_2018['sat_participation_2017'] > 80)][0:50]
participation_sat_2017[['state','sat_participation_2017','sat_total_2017', 'sat_participation_2018', 'sat_total_2018']]

In [ ]:
#colorado had a 70% drop in particiation 
# states with free act testing made the list, Teenese and Minnesota dont offer free ACT  but still ranked top 15.

participation_act_2017 = df_2017_2018[(df_2017_2018['act_participation_2017'] > 99)][0:50]
participation_act_2017[['state','act_participation_2017','act_composite_2017', 'act_participation_2018','act_composite_2018']]

## Visualize the data

There's not a magic bullet recommendation for the right number of plots to understand a given dataset, but visualizing your data is *always* a good idea. Not only does it allow you to quickly convey your findings (even if you have a non-technical audience), it will often reveal trends in your data that escaped you when you were looking only at numbers.

Some recommendations on plotting:
- Plots have titles
- Plots have axis labels
- Plots have appropriate tick labels
- All text is legible in a plot
- Plots demonstrate meaningful and valid relationships
- Plots are interpreted to aid understanding

There is such a thing as too many plots, and there are a *lot* of bad plots. You might make some! (But hopefully not with the guided prompts below).

#### Use Seaborn's heatmap with pandas `.corr()` to visualize correlations between all numeric features

Heatmaps are generally not appropriate for presentations, and should often be excluded from reports as they can be visually overwhelming. **However**, they can be extremely useful in identify relationships of potential interest (as well as identifying potential collinearity before modeling).

*example*:
```python
sns.heatmap(df.corr())
```

Please take time to format your output, adding a title. Look through some of the additional arguments and options. (Axis labels aren't really necessary, as long as the title is informative).

In [ ]:

plt.figure(figsize=(20,12))
plt.title('Relationships of Potential Interest', fontsize=24)
#sns.dark_palette("palegreen", as_cmap=True)
corr = df_2017_2018.corr()
mask = np.zeros_like(corr, dtype=np.bool)
mask[np.triu_indices_from(mask)] = True
sns.heatmap(corr,mask=mask, cmap = sns.diverging_palette(240, 10, n = 13, as_cmap=True), center = True, annot = True,
           vmin=-1, vmax=1);

#### Define a custom function to subplot histograms

We have data for two tests for two years. We only have composite (and not subtest scores) for the 2018 ACT. We should write a function that will take the names of 2+ columns and subplot histograms. While you can use pandas plotting or Seaborn here, matplotlib gives you greater control over all aspects of your plots.

[Helpful Link for Plotting Multiple Figures](https://matplotlib.org/users/pyplot_tutorial.html#working-with-multiple-figures-and-axes)

Here's some starter code:

In [ ]:
def subplot_histograms(dataframe, list_of_columns, list_of_titles, list_of_xlabels):
    nrows = int(np.ceil(len(list_of_columns)/2)) # Makes sure you have enough rows
    fig, ax = plt.subplots(nrows=nrows, ncols=2, figsize= (12,40))
    ax = ax.ravel() # Ravel turns a matrix into a vector, which is easier to iterate
    
    for i, column in enumerate(list_of_columns): # Gives us an index value to get into all our lists
        ax[i].hist(dataframe[column]) # feel free to add more settings
        ax[i].set_title(list_of_titles[i])
    fig.tight_layout()    
        # Set titles, labels, etc here for each subplot

In [ ]:
subplot_histograms(df_2017_2018, ['sat_participation_2017', 'sat_reading_2017', 'sat_math_2017',
                   'sat_total_2017', 'act_participation_2017', 'act_english_2017',
                   'act_math_2017', 'act_reading_2017', 'act_science_2017',
                   'act_composite_2017', 'sat_participation_2018', 'sat_reading_2018',
                   'sat_math_2018', 'sat_total_2018', 'act_participation_2018',
                   'act_composite_2018'],
                   ['SAT Participation 2017', 'SAT Reading 2017', 'SAT Math 2017', 'SAT Total 2017', 'ACT Participation 2017', 
                    'ACT English 2017',' ACT Math 2017', 'ACT reading 2017', 'ACT Science 2017', 'ACT Composite 2017',
                   'SAT Participation 2018', 'SAT Reading 2018', 'SAT Math 2018', 'Sat Total 2018','ACT Participation 2018', 'ACT Composite 2018'],
                   [])



In [ ]:
df_2017_2018.columns

In [ ]:
# next time Kiddo !!
# df_2017_2018[['sat_participation_2017', 'sat_reading_2017', 'sat_math_2017',
#        'sat_total_2017', 'act_participation_2017', 'act_english_2017',
#        'act_math_2017', 'act_reading_2017', 'act_science_2017',
#        'act_composite_2017', 'sat_participation_2018', 'sat_reading_2018',
#        'sat_math_2018', 'sat_total_2018', 'act_participation_2018',
#        'act_composite_2018']].hist();
# plt.figure(figsize=(1000,1000))
# plt.show();

In [ ]:
sns.pairplot(df_2017_2018[['sat_math_2017','sat_math_2018']]);


In [ ]:
#another fuck up
# df_2017_2018.iloc[:,1:].hist()
# plt.figure(figsize=(10,10))
# plt.show();
df_2017_2018_agg.plot(kind="line", figsize=(10, 10));

#### Plot and interpret histograms 
For each of the following:
- Participation rates for SAT & ACT
- Math scores for SAT & ACT
- Reading/verbal scores for SAT & ACT

In [ ]:
# df_2017_2018['sat_participation_2017'].hist();

#Participation rates for SAT & ACT
fig, ax = plt.subplots(nrows = 2, ncols=2, figsize=(10,10))
df_2017_2018['sat_participation_2017'].plot(kind='hist', ax= ax[0,0], title='SAT Participation 2017')
df_2017_2018['sat_participation_2018'].plot(kind='hist', ax= ax[0,1], title='SAT Participation 2018')
df_2017_2018['act_participation_2017'].plot(kind='hist', ax= ax[1,0], title='ACT Participation 2017')
df_2017_2018['act_participation_2018'].plot(kind='hist', ax= ax[1,1], title='ACT Participation 2018')
fig.tight_layout()

In [ ]:
#Math scores for SAT & ACT
fig, ax = plt.subplots(nrows = 2, ncols=2, figsize=(10,10))
df_2017_2018['sat_math_2017'].plot(kind='hist', ax= ax[0,0], title='SAT Math 2017')
df_2017_2018['sat_math_2018'].plot(kind='hist', ax= ax[0,1], title='SAT Math 2018')
df_2017_2018['act_math_2017'].plot(kind='hist', ax= ax[1,0], title='ACT Math 2017')
#df_2017_2018['act_math_2018'].plot(kind='hist', ax= ax[1,1], title='ACT Math 2018')
fig.tight_layout()

In [ ]:
#Reading/verbal scores for SAT & ACT
fig, ax = plt.subplots(nrows = 2, ncols=2, figsize=(10,10))
df_2017_2018['sat_reading_2017'].plot(kind='hist', ax= ax[0,0], title='SAT Reading 2017')
df_2017_2018['sat_reading_2018'].plot(kind='hist', ax= ax[0,1], title='SAT Reading 2018')
df_2017_2018['act_reading_2017'].plot(kind='hist', ax= ax[1,0], title='ACT Reading 2017')
df_2017_2018['act_english_2017'].plot(kind='hist', ax= ax[1,1], title='ACT English 2017')
fig.tight_layout()

# In [41]:
# SAT_ACT.hist(column='Reading_ACT', xlabelsize=30, ylabelsize=30, figsize=(20,10), bins=20)
# plt.xlabel('ACT Verbal Scores', fontsize=20)
# plt.ylabel('Number of States', fontsize=20)
# plt.title('ACT Verbal Reading Scores Across the 50 States')
# plt.show

In [ ]:
df_2017_2018.head(3)

In [ ]:
df_2017_2018.groupby('state')['sat_participation_2017'].mean().plot(kind="line", figsize=(15,10));
# just fucking aorund here

#### Plot and interpret scatter plots

For each of the following:
- SAT vs. ACT math scores for 2017
- SAT vs. ACT verbal/reading scores for 2017
- SAT vs. ACT total/composite scores for 2017
- Total scores for SAT 2017 vs. 2018
- Composite scores for ACT 2017 vs. 2018

Plot the two variables against each other using matplotlib or Seaborn

Your plots should show:
- Two clearly labeled axes
- A proper title
- Using colors and symbols that are clear and unmistakable

**Feel free to write a custom function, and subplot if you'd like.** Functions save both time and space.


In [ ]:
# comparing two non-time quantitavie variables
# I see that outlier at iloc[20,3]
                   
df_2017_2018.plot.scatter(x='sat_participation_2017',
                          y='sat_math_2017',
                          c='red', title='SAT Math Scores v. Participation')
fig.tight_layout()

In [ ]:
df_2017_2018.plot.scatter('sat_reading_2017', 'sat_participation_2017')
plt.xlabel('sat reading 2017', fontsize=15)
plt.ylabel('sat participation 2017', fontsize=15)
plt.title('SAT reading v. Participation');

In [ ]:
#SAT vs. ACT total/composite scores for 2017

df_2017_2018.plot.scatter('sat_total_2017', 'act_composite_2017', c='r')
plt.xlabel('sat_total_2017', fontsize=15)
plt.ylabel('act_composite_2017', fontsize=15)
plt.title('SAT Total v. ACT Composite');

In [ ]:
#Total scores for SAT 2017 vs. 2018
cat=fig.gca()

cat = df_2017_2018.plot(kind='scatter', x='sat_total_2017', y='sat_participation_2017', color='darkgreen',
                 label='SAT 2017 Group', title= 'Score v. Participaion 2017-2018')


# df_2017_2018.plot(kind='scatter', x='sat_total_2018', y='sat_participation_2018', color='red',
#                  label='SAT 2018 Group', ax= cat)
# plt.show()

In [ ]:
df_2017_2018.plot(kind='scatter', x='sat_total_2018', y='sat_participation_2018', color='red',
                 label='SAT 2018 Group')


In [ ]:
#Total scores for SAT 2017 vs. 2018
cat=fig.gca()

cat = df_2017_2018.plot(kind='scatter', x='sat_total_2017', y='sat_participation_2017', color='darkgreen',
                 label='SAT 2017 Group', title= 'Score v. Participaion 2017-2018')

df_2017_2018.plot(kind='scatter', x='sat_total_2018', y='sat_participation_2018', color='red',
                 label='SAT 2018 Group', ax= cat)
plt.show()

In [ ]:
participation_sat_2017.plot(kind='scatter', x='sat_participation_2017', y='sat_total_2017', color='blue',
                  label='SAT 2017 Group', title= 'States with SAT participation rate over 80% ');


In [ ]:
vax = participation_act_2017.plot(kind='scatter', x='act_participation_2017', y='act_composite_2017', color='orange',
                 label='ACT 2017 Group', title= 'States with ACT participation rate over 90%');

In [ ]:
vax=fig.gca()

vax =participation_sat_2017.plot(kind='scatter', x='sat_participation_2017', y='sat_total_2017', color='blue',
                  label='SAT 2017 Group', title= 'States with SAT participation rate over 80% ')
participation_act_2017.plot(kind='scatter', x='act_participation_2017', y='act_composite_2017', color='orange',
                 label='ACT 2017 Group', title= 'States with ACT participation rate over 90%', ax=vax);

#### Plot and interpret boxplots

For each numeric variable in the dataframe create a boxplot using Seaborn. Boxplots demonstrate central tendency and spread in variables. In a certain sense, these are somewhat redundant with histograms, but you may be better able to identify clear outliers or differences in IQR, etc.

Multiple values can be plotted to a single boxplot as long as they are of the same relative scale (meaning they have similar min/max values).

Each boxplot should:
- Only include variables of a similar scale
- Have clear labels for each variable
- Have appropriate titles and labels

In [ ]:
# Code
#
fig = plt.figure(figsize=(15,6))
ax = fig.gca()

ax = sns.boxplot(df_2017_2018.act_science_2017, orient='h',
                fliersize=8, linewidth=1.5, notch=False,
                saturation=0.5, ax=ax)

ax.set_ylabel('ACT science 2017', fontsize=16)
ax.set_title('ACT science scores 2017\n', fontsize=20)

plt.show()

In [ ]:
#shows me the outlier, fucking maryland

fig = plt.figure(figsize=(5,5))
sns.stripplot(x="act_science_2017", y=df_2017_2018.iloc[18:22,0], data=df_2017_2018);

In [ ]:
fig = plt.figure(figsize=(15,6))
ax = fig.gca()

ax = sns.boxplot(df_2017_2018.sat_math_2017, orient='h',
                fliersize=8, linewidth=1.5, notch=False,
                saturation=0.5, ax=ax)

ax.set_ylabel('sat math 2017', fontsize=16)
ax.set_title('SAT scores 2017\n', fontsize=20)

plt.show()

In [ ]:
#Draw a boxplot with nested grouping by two categorical variables; not very useful... maybe change the variables
fig= plt.figure(figsize=(30,15))
ax= fig.gca()
ax= sns.boxplot(x="sat_participation_2018", y="sat_math_2018", color="blue",
                  data=df_2017_2018, palette="Set3", ax=ax)
ax.set_title('SAT math scores')
plt.show()

#### Feel free to do additional plots below
*(do research and choose your own chart types & variables)*

Are there any additional trends or relationships you haven't explored? Was there something interesting you saw that you'd like to dive further into? It's likely that there are a few more plots you might want to generate to support your narrative and recommendations that you are building toward. **As always, make sure you're interpreting your plots as you go**.

In [ ]:
fig = plt.figure(figsize=(12,12))
ax = fig.gca()

ax = sns.boxplot(data=df_2017_2018, orient='h', fliersize=10, linewidth=3, notch=True,
                 saturation=0.5, ax=ax)

ax.set_title('All variables boxplot\n')
plt.show()

#### (Optional): Using Tableau, create a choropleth map for each variable using a map of the US. 

Save this plot as an image file in an images directory, provide a relative path, and insert the image into notebook in markdown.

## Descriptive and Inferential Statistics

#### Summarizing Distributions

Above, we used pandas `describe` to provide quick summary statistics of our numeric columns. We also demonstrated many visual relationships.

As data scientists, having a complete understanding of data is imperative prior to modeling.

While we will continue to build our analytic tools, we know that measures of *central tendency*, *spread*, and *shape/skewness* provide a quick summary of distributions.

For each variable in your data, summarize the underlying distributions (in words & statistics)
 - Be thorough in your verbal description of these distributions.
 - Be sure to back up these summaries with statistics.


Answers: This Exploratory Data Analysis were based on 17 different variables, see data frame below. There were some missing variables that would help us determine a deeper interpretation. The data for the ACT test in 2018 is missing data on the 4 sections of the test: Math, Science, English, Reading. 

SAT participation in 2017 has a histogram with a skew to the right and you can easily see this in the mean v. median. When the mean is greater than the median you will get a histogram to the right. The standard deviation is .035, which is a large deviation, and thats why you see such a large number on participation plotted farther away from the mean. 

SAT reading 2017 has a median larger than the mean which should give us a histogram skewed to the left. However, when you plot this graph it shows like its skewed to the right and I believe the data skewed to the right because of the high standard deviation of 45.

SAT Math 2017 have a mean and a median that are very close. This proximity would normally suggest that this data is evenly distributed. However, when this histogram is plotted it gives us a graph skewed to the right. The standard deviation for this data frame is at 84.

SAT total 2017 has the mean greater than the median, when interpreted just by those two numbers you can assume that this graph will skew to the right. When we plot the histogram we do see that skew to the right. 

ACT participation 2017 has a median that is greater than the mean which should plot a histogram skewed to the left. When we see the actual plot we do get that skew to the left and that is mainly due to some states having 100% participation rates.

ACT english 2017 has a median that is really close to the mean. This relationship will plot a histogram that is evenly distributed. The standard deviation here is really low, so thats another clue that will tell us the distribution will be normal.

ACT math 2017 has a median that is only .2 points away. This relationship between the mean and median, along with the low standard deviation of 2.3, will plot a normal distribution.

ACT reading 2017 has a mean greater than the median, but only by .2 points. Again, we will see a histogram with a normal distribution, considered the standard deviation of 2.06.

ACT science 2017 the mean and media are only .2 points from one another. The .2 difference will tell us that the histograph will have a normal distribution, backed by the 3.2 standard deviation. 

ACT composite 2017 has a mean pretty much equal to its median. This is the collection of all the variables of ACT so it would make sense that this would plot as a normal distribution histograph. If you need further proof look at the standard deviation of 2.02

SAT participation 2018 has higher mean, median, and standard deviation from 2017. This distribution is skewed to the right and thats mainly to the lower participating states. This data does have a high standard deviation of 37, which is just over 2 points from 2017. So the YoY increase in participation is dismal.

SAT reading 2018  has a mean greater than the median, by about 13 points. Include the high standard deviation of 45 and we see a histograph skewed to the right. 

SAT math 2018 has a mean greater than the median with a standard deviation of 49.5. When we look at the histograph and see that a lot of the data plots along the mean of all the scores for 2018. 

SAT total 2018 total is skewed to the right. We have a mean that is greater than the median. Some outliers are seen here that come from the performances from certain states.

ACT participation 2018 is skewed to the left. The median is greater than the mean. This median comes from 13 states having a 100% participation rate. 

ACT composite 2018 has all the variables to plot a nor distributed histogram, but when we look the graph it looks like it has two tails. The range of scores is so small that it throws the graph of from what the mean, median, and standard deviation conclude. 

In [ ]:
# all the stats for the descriptive summary above
df_2017_2018_agg.T

#### We generally assuming that data we sample from a population will be normally distributed. Do we observe this trend?

Answer: No, most charts show that individual score and participation rates are skewed. The only numbers that are normally didstributed are the totals/composite values. 

Does This Assumption Hold for:
    - Math
    - Reading
    - Rates
Explain your answers for each distribution and how you think this will affect estimates made from these data.

Answer: 
Math- SAT has the data that show a normal distribution. ACT math 2018 holds a normal distribution. 
Reading- We dont see a normal distribution for either test. 
Rates- skewed and I believe it all depends on how the states handle these tests.

#### Estimate Limits of Data

Suppose we only seek to understand the relationship between SAT and ACT participation rates in 2017. 

##### Does it make sense to conduct statistical inference given these data specifically? 

Why or why not?

*(think about granularity, aggregation, the relationships between populations size & rates...consider the actually populations these data describe in answering this question)*

I would think that it would could use statistical inference to find something like a confidnece interval. If you take a sample from the math scores, science scores, or the total score this would show a good representation of the population. We could arrive to different conclusion about population parameters based on samples from each of these variables. 

Answer:

##### Is it appropriate to compare *these* specific SAT and ACT math scores? 

Why or why not?

Statistical Inference wouldnt work on these two data sets because they dont follow the Gaussian curve, in which case parametric tests are applied.  I see these scores as unpaired becuase SAT math scores and ACT math scores have no possibility of one influencing the other. 

#### Statistical Evaluation of Distributions 

**If you feel it's appropriate**, using methods we discussed in class, run hypothesis tests to compare variables of interest in our dataset. 

In [ ]:
# Code:
sat_part_mean = np.mean(df_2017_2018['sat_participation_2017'])
act_part_mean = np.mean(df_2017_2018['act_participation_2017'])
sat_part_mean

In [ ]:
act_part_mean

In [ ]:
sat_data = df_2017_2018['sat_participation_2017'][0:50]
act_data = df_2017_2018['act_participation_2017'][0:50]

In [ ]:
stats.ttest_ind(sat_data, act_data)

If the p-value is less than or equal to the alpha (p< .05), then we reject the null hypothesis, and we say the result is statistically significant. 
The p value is less than alpha, we fail to reject the null hypothesis, so we reject the null hypotheses 

## Outside Research

Based upon your observations, choose **three** states that demonstrate interesting trends in their SAT and/or ACT participation rates. Spend some time doing outside research on state policies that might influence these rates, and summarize your findings below. **Feel free to go back and create new plots that highlight these states of interest**. If you bring in any outside tables or charts, make sure you are explicit about having borrowed them. If you quote any text, make sure that it renders as being quoted. (Make sure that you cite your sources -- check with you local instructor for citation preferences).

In [ ]:
#colorado had a 70% drop in particiation 
# states with free act testing made the list, Teenese and Minnesota dont offer free ACT  but still ranked top 15.
#Do any states with 100% participation on a given test have a rate change year-to-year?
#D.C had an 8% drop YoY
#Florida had the highest sat_participation absolute change YoY among all the states that had 80%+
#Colorado offers free SAT but didnt even get an 80% participation rate. 
#In 2017 thier participation rate was 11% in 2017 then increased to 28%
# 8 states offer Free SAT
#16 states 
cat=fig.gca()

cat = df_2017_2018.plot(kind='scatter', x='sat_total_2017', y='sat_participation_2017', color='darkgreen',
                 label='SAT 2017 Group', title= 'Score v. Participaion 2017-2018')

df_2017_2018.plot(kind='scatter', x='sat_total_2018', y='sat_participation_2018', color='red',
                 label='SAT 2018 Group', ax= cat)
plt.show()

## Conclusions and Recommendations

Based on your exploration of the data, what are you key takeaways and recommendations? Choose one state with a lower participation rate and provide a suggestion for how the College Board might increase participation amongst graduating seniors in this state. Are there additional data you desire that would better inform your investigations?

I reccomned that we could really increase participation rates by making the ACT the test requeired for admissions. Michigan is a perfect example. Professor Hyman from the Unviserity of Michigan studied the effects of 'free ACT testing', he did a study tracking this data. Michigan began requiring public school juniors to take the ACT in 2007, and the share of high school graduates taking a college entrance exam rose immediately to nearly 99 percent from 54 percent. That growth was even sharper among low-income students; only 35 percent had been taking the test. Professor Hyman also found for every 1,000 students who took a college exam when it was optional, and scored high enough to attend a selective college, another 230 high scorers appeared once the test was mandatory. For low-income students, the effect was larger: For every 1,000 students who scored well on the optional test, an additional 480 did so on the mandatory test. Jumpforward ten years and you see that Michigan's ACT numbers composite scores are among the highest in the nation and score 3 points above the mean for either years. 


In [ ]:
df_2017_2018.iloc[22][['state','act_composite_2017','act_composite_2018']]


In [ ]:
df_2017_2018[['state','act_composite_2017','act_composite_2018']].mean()
